### ISSUE TITLE
google.appengine.tools.devappserver2.wsgi_server.BindError: Unable to bind localhost:8080

### ISSUE THREAD

**Issue by elharo** [GoogleCloudPlatform/google-cloud-eclipse#393]
May go into dev_apperver, but when we can't bind to the expected port, we should probably offer a better error message. 

Alternatively, we shouldn't try to bind to 8080 at all and instead pick a known available port.

**Comment by chanseokoh** [comment]
Since we allow configuring a port, I think we should fail starting the server when the port is in use instead of trying to pick a free port.

About the error message, it would be nice to have a custom dialog, but I remember @briandealwis recommends conforming to the way the WTP framework handles errors. I am thinking of throwing a `CoreException` before starting the dev server (in `LocalAppEngineServerBehaviour`) by checking the specified port.

**Comment by briandealwis** [comment]
The same problem happens with the API port and the admin port.  I wonder if dev_appserver returns a non-0 exit code, and if we can pick up on that?

**Comment by chanseokoh** [comment]
> The same problem happens with the API port and the admin port.

Good point. For the service port, I think it makes sense to fail if the port is in use, since our model is to let users explicitly configure it.

For the admin port, at the moment, should we let dev_appserver always pick a free port for the admin port? Or, let it pick a free port only when the default 8000 is in use? (Ultimately, we should allow configuring the admin port and make the behavior same as the service port when the port is in use, but that'd be the next time.)

For the API port, looks like dev_appserver always picks a free port, so I think we are good.

> I wonder if dev_appserver returns a non-0 exit code, and if we can pick up on that?

dev_appserver may return a non-0 exit code, but to actually know if it was because of port binding failure, we will probably have to parse the console output. (I think the server doesn't provide formatted output in this case, since the error output is due to an uncaught Python exception.) Considering this, I thought it's better to check ports ourselves before starting the server.

**Comment by briandealwis** [comment]
> Or, let it pick a free port only when the default 8000 is in use? 

I think that makes sense.

WTP provides a helper to determine if a port is free: `org.eclipse.wst.server.core.util.SocketUtil#isPortInUse()`

**Comment by elharo** [comment]
Be wary of overconfiguration. Every switch, toggle, and button we add to the UI increases cognitive load more than linearly (and conversely every one we remove decreases cognitive load). I'd prefer to wait to support explicit choice of the admin and API ports until we know a significant number of users or use cases actually require this.  Right now I suspect this is YAGNI.

**Comment by patflynn** [comment]
FYI in IJ we allow the user to configure the service, admin, and API
ports.  I expect that by and large the service and admin ports are the only
ones they are going to care to configure.
